src/persist-client/src/internal/state_diff.rs

```diff
@@ -865,9 +865,12 @@ fn apply_compaction_lenient<'a, T: Timestamp + Lattice>(
 }
 
 impl ProtoStateFieldDiffs {
-    pub fn push_data(&mut self, mut data: Vec<u8>) {
-        self.data_lens.push(u64::cast_from(data.len()));
-        self.data_bytes.append(&mut data);
+    pub fn encode_proto<M: prost::Message>(&mut self, msg: &M) {
+        let encoded_length = msg.encoded_len();
+        self.data_lens.push(u64::cast_from(encoded_length));
```

Comment:
it stresses me out a decent bit to rely on `encoded_len` in this way. it's indeed used in the default impl of `prost::Message::encode` to allocate a buffer that is guaranteed to have enough space, but still I think I'd be much more comfortable if it was structured more like the following:

```rust
self.data_bytes.reserve(msg.encoded_len());
let len_before = self.data_bytes.len();
msg.encode(&mut self.data_bytes).expect("reserved enough space");
self.data_lens.push(self.data_bytes-len_before);
```

also noticed while I was in here that the new impl is calling `encoding_len()` twice, once here explicitly and once in inside `msg.encode`. while you have the benchmarks and everything sitting around, would to mind running an experiment for me to switch from `prost::Message::encode` to `prost::Message::encode_raw` and leaving the results in a comment inline here in the code, please?

Reply:
Totally makes sense with using `encoded_len`, I updated the PR to track the exact bytes written.

My machine is definitely noisy, but when using `encode_raw(...)` I did see a _slight_ runtime improvement, and no difference in the number of allocations. Because all we do anyways is panic on a failure of `encode(...)`, I swapped to use `encode_raw(...)`, so we have the same behavior, but better perf.